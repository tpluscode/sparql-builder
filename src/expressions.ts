export { IN } from './lib/expressions/in'
export { VALUES } from './lib/expressions/values'
