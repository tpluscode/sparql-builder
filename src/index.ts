export { sparql } from '@tpluscode/rdf-string'
export { DELETE } from './lib/DeleteBuilder.js'
export type { DeleteInsertQuery as DeleteInsert, DeleteData } from './lib/DeleteBuilder.js'
export { INSERT } from './lib/InsertBuilder.js'
export type { InsertQuery as Insert, InsertData } from './lib/InsertBuilder.js'
export { ASK } from './lib/AskBuilder.js'
export type { AskQuery as Ask } from './lib/AskBuilder.js'
export { DESCRIBE } from './lib/DescribeBuilder.js'
export type { DescribeQuery as Describe } from './lib/DescribeBuilder.js'
export { CONSTRUCT } from './lib/ConstructBuilder.js'
export type { ConstructQuery as Construct } from './lib/ConstructBuilder.js'
export { SELECT } from './lib/SelectBuilder.js'
export type { SelectQuery as Select } from './lib/SelectBuilder.js'
export { WITH } from './lib/WithBuilder.js'
export { prefixes } from '@tpluscode/rdf-string'
export type { SparqlTemplateResult } from '@tpluscode/rdf-string'
export type {
  SparqlAskExecutable, SparqlGraphQueryExecutable, SparqlQueryExecutable, SparqlUpdateExecutable,
} from './lib/index.js'
