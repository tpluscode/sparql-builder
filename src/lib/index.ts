import { AskQuery, ConstructQuery, QueryOptions, SelectQuery, UpdateQuery } from 'sparql-http-client'
import { SparqlTemplateResult } from '@tpluscode/rdf-string'
import prologue, { PrologueBuilder } from './partials/prologue'

interface SparqlBuildOptions {
  base?: string
}

export type SparqlExecuteOptions = QueryOptions & SparqlBuildOptions

export interface SparqlQuery extends PrologueBuilder {
  build(options?: SparqlBuildOptions): string
  _getTemplateResult(): SparqlTemplateResult
}

export interface SparqlQueryExecutable {
  execute<TQuery extends SelectQuery<any>>(client: TQuery, requestInit?: SparqlExecuteOptions): ReturnType<TQuery['select']>
}

export interface SparqlGraphQueryExecutable {
  execute<TQuery extends ConstructQuery<any>>(client: TQuery, requestInit?: SparqlExecuteOptions): ReturnType<TQuery['construct']>
}

export interface SparqlUpdateExecutable {
  execute<TQuery extends UpdateQuery<any>>(client: TQuery, requestInit?: SparqlExecuteOptions): ReturnType<TQuery['update']>
}

export interface SparqlAskExecutable {
  execute<TQuery extends AskQuery<any>>(client: TQuery, requestInit?: SparqlExecuteOptions): ReturnType<TQuery['ask']>
}

type TBuilder = Pick<SparqlQuery, 'build'> & Pick<SparqlTemplateResult, '_toPartialString'>

// eslint-disable-next-line no-unused-vars
export default function Builder<T extends SparqlQuery>(): TBuilder & T {
  return {
    ...prologue(),
    build(this: SparqlQuery, { base }: SparqlBuildOptions = {}): string {
      const queryResult = this._getTemplateResult().toString({
        base,
      })

      if (this.prologueResult) {
        return `${this.prologueResult}\n\n${queryResult}`
      }

      return queryResult
    },
    _toPartialString(this: SparqlQuery, options: any) {
      return this._getTemplateResult()._toPartialString(options)
    },
  } as any
}
