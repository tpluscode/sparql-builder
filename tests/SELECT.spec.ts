import namespace from '@rdfjs/namespace'
import RDF from '@rdfjs/data-model'
import { SELECT } from '../src'
import { sparqlClient } from './_mocks'

describe('SELECT', () => {
  it('executes as select', () => {
    // given
    const client = sparqlClient()

    // when
    SELECT``.execute(client)

    // then
    expect(client.select).toHaveBeenCalled()
  })

  it('creates a simple select/where', () => {
    // given
    const expected = 'SELECT ?s ?p ?o WHERE { ?s ?p ?o }'

    // when
    const actual = SELECT`?s ?p ?o`.WHERE`?s ?p ?o`.build()

    // then
    expect(actual).toMatchQuery(expected)
  })

  it('combines multiple WHERE clauses', () => {
    // given
    const expected = 'SELECT * WHERE { ?s ?p ?o. ?a ?b ?c }'

    // when
    const actual = SELECT`*`
      .WHERE`?s ?p ?o .`
      .WHERE`?a ?b ?c .`
      .build()

    // then
    expect(actual).toMatchQuery(expected)
  })

  it('adds FROM when default graph set', () => {
    // given
    const expected = 'SELECT * FROM <urn:foo:bar> WHERE { ?s ?p ?o }'

    // when
    const actual = SELECT`*`.FROM(RDF.namedNode('urn:foo:bar')).WHERE`?s ?p ?o`.build()

    // then
    expect(actual).toMatchQuery(expected)
  })

  it('does not add FROM when graph is defaultGraph', () => {
    // given
    const expected = 'SELECT * WHERE { ?s ?p ?o }'

    // when
    const actual = SELECT`*`.FROM(RDF.defaultGraph()).WHERE`?s ?p ?o`.build()

    // then
    expect(actual).toMatchQuery(expected)
  })

  it('supports LIMIT/OFFSET', () => {
    // given
    const expected = 'SELECT ?s ?p ?o WHERE { ?s ?p ?o } LIMIT 15 OFFSET 40'

    // when
    const actual = SELECT`?s ?p ?o`.WHERE`?s ?p ?o`.LIMIT(15).OFFSET(40).build()

    // then
    expect(actual).toMatchQuery(expected)
  })

  it('can be constructed with a base', () => {
    // given
    const ns = namespace('http://example.com/')
    const expected = `
     BASE <http://example.com/>
     
     SELECT *
     FROM <graph>
     WHERE {
      <person> a <Person> 
     }`

    // when
    const actual = SELECT`*`
      .FROM(ns.graph)
      .WHERE`${ns.person} a ${ns.Person}`
      .build({
        base: 'http://example.com/',
      })

    // then
    expect(actual).toMatchQuery(expected)
  })

  it('can be ordered by variable', () => {
    // given
    const expected = 'SELECT ?s ?p ?o WHERE { ?s ?p ?o } ORDER BY ?s'

    // when
    const s = RDF.variable('s')
    const actual = SELECT`${s} ?p ?o`
      .WHERE`${s} ?p ?o`
      .ORDER().BY(s)
      .build()

    // then
    expect(actual).toMatchQuery(expected)
  })

  it('can be ordered by desc(?variable)', () => {
    // given
    const expected = 'SELECT ?s ?p ?o WHERE { ?s ?p ?o } ORDER BY desc(?s)'

    // when
    const s = RDF.variable('s')
    const actual = SELECT`${s} ?p ?o`
      .WHERE`${s} ?p ?o`
      .ORDER().BY(s, true)
      .build()

    // then
    expect(actual).toMatchQuery(expected)
  })

  it('can be ordered by multiple variables', () => {
    // given
    const expected = 'SELECT ?s ?p ?o WHERE { ?s ?p ?o } ORDER BY desc(?s) ?o'

    // when
    const s = RDF.variable('s')
    const o = RDF.variable('o')
    const actual = SELECT`${s} ?p ${o}`
      .WHERE`${s} ?p ${o}`
      .ORDER().BY(s, true).THEN.BY(o)
      .build()

    // then
    expect(actual).toMatchQuery(expected)
  })

  it('can be ordered and limited, when calls are reversed', () => {
    // given
    const expected = 'SELECT ?s ?p ?o WHERE { ?s ?p ?o } ORDER BY ?s LIMIT 20'

    // when
    const s = RDF.variable('s')
    const actual = SELECT`${s} ?p ?o`
      .WHERE`${s} ?p ?o`
      .LIMIT(20)
      .ORDER().BY(s)
      .build()

    // then
    expect(actual).toMatchQuery(expected)
  })

  describe('DISTINCT', () => {
    it('creates correct SPARQL', () => {
      // given
      const expected = 'SELECT DISTINCT * WHERE { ?s ?p ?o }'

      // when
      const actual = SELECT.DISTINCT`*`.WHERE`?s ?p ?o`.build()

      // then
      expect(actual).toMatchQuery(expected)
    })
  })

  describe('REDUCED', () => {
    it('creates correct SPARQL', () => {
      // given
      const expected = 'SELECT REDUCED * WHERE { ?s ?p ?o }'

      // when
      const actual = SELECT.REDUCED`*`.WHERE`?s ?p ?o`.build()

      // then
      expect(actual).toMatchQuery(expected)
    })
  })
})
