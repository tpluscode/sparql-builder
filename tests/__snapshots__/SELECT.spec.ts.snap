// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SELECT can have additional prologue 1`] = `
"#pragma join.hash off
BASE <http://foo.bar/baz>

SELECT ?s ?p ?o

WHERE {
?s ?p ?o
}



"
`;
